### PR TITLE
fix mypy and ruff errors

### PR DESCRIPTION
**Pull request by CunliangGeng**
This PR tries to solve all type hints errors (mypy), formatting errors (ruff) and typos for refactored code. As for non-refactored code, we keep it as it is.

### As for type hints, the changes follow the rules below:
- use `@overload` for different data types of the same argument
- use broader type hints `Sequence` and `Mapping` to replace `list` and `dict`, respectively, as much as possible
- Use `TYPE_CHECKING` to only avoid circular imports

### CODE REVIEW

**Comment by CunliangGeng** [comment]
* **#260** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/260?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#259** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/259?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#258** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/258?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#257** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/257?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#256** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/256?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#255** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/255?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#254** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/254?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#253** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/253?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#252** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/252?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#250** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/250?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#249** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/249?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#248** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/248?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#247** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/247?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#246** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/246?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#245** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/245?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `dev`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @CunliangGeng and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by CunliangGeng** [comment]
It's ready for review.

**Comment by CunliangGeng** [comment]
It's ready for review.

File: src/nplinker/genomics/antismash/antismash_loader.py [hunk 1]
```diff
@@ -97,7 +98,7 @@ def get_bgcs(self) -> list[BGC]:
         return self._bgcs
 
     @staticmethod
-    def _parse_bgcs(bgc_files: dict[str, str]) -> list[BGC]:
+    def _parse_bgcs(bgc_files: Mapping[str, str]) -> list[BGC]:
```
Comment (gcroci2) [inline src/nplinker/genomics/antismash/antismash_loader.py:101]: Why here not using `Sequence[BGC]` for the returned type?

File: src/nplinker/genomics/mibig/mibig_loader.py [hunk 2]
```diff
@@ -58,15 +58,15 @@ def parse_data_dir(data_dir: str) -> dict[str, str]:
             file_dict[fname] = file
         return file_dict
 
-    def get_metadatas(self) -> dict[str, MibigMetadata]:
+    def get_metadata(self) -> dict[str, MibigMetadata]:
```
Comment (gcroci2) [inline src/nplinker/genomics/mibig/mibig_loader.py:61]: Why here not using `Mapping`?

File: src/nplinker/genomics/mibig/mibig_loader.py [hunk 3]
```diff
@@ -58,15 +58,15 @@ def parse_data_dir(data_dir: str) -> dict[str, str]:
...
         """Get MibigMetadata objects.
 
         Returns:
             The key is BGC accession (file name) and the value is MibigMetadata object
         """
         return self._metadata_dict
 
-    def _parse_metadatas(self) -> dict[str, MibigMetadata]:
+    def _parse_metadata(self) -> dict[str, MibigMetadata]:
```
Comment (gcroci2) [inline src/nplinker/genomics/mibig/mibig_loader.py:69]: Same as above, why not using `Mapping`?

File: src/nplinker/genomics/utils.py [hunk 4]
```diff
@@ -65,7 +67,9 @@ def generate_mappings_genome_id_bgc_id(
     logger.info("Generated genome-BGC mappings file: %s", output_file)
 
 
-def add_strain_to_bgc(strains: StrainCollection, bgcs: list[BGC]) -> tuple[list[BGC], list[BGC]]:
+def add_strain_to_bgc(
+    strains: StrainCollection, bgcs: Sequence[BGC]
+) -> tuple[list[BGC], list[BGC]]:
```
Comment (gcroci2) [inline src/nplinker/genomics/utils.py:72]: Why not using `Sequence` for the returned type?

**Review by gcroci2 — COMMENTED** [review]
In general, there are many lines touched by this PR where `Mapping` or `Sequence` are not used for type hints instead of `dict` and `list`. I started commenting on them, but since I realized they are quite a lot I wonder if there is a reason for this. Before I keep leaving many comments, better to know ;)

File: src/nplinker/genomics/antismash/antismash_loader.py (hunk 1 above)
Comment (CunliangGeng) [reply src/nplinker/genomics/antismash/antismash_loader.py:101]: The principle is keeping robust for input (arguments) and keeping specific for output.
That is why it's nice to use wider type for input values but specific type for returned values.

File: src/nplinker/genomics/abc.py [hunk 5]
```diff
@@ -18,7 +19,7 @@ def __init__(self, data_dir: str):
         self.data_dir = data_dir
 
     @abstractmethod
-    def get_files(self) -> dict[str, str]:
+    def get_files(self) -> Mapping[str, str]:
```
Comment (gcroci2) [inline src/nplinker/genomics/abc.py]: Maybe here better to be specific then? No big difference of course, but it's more coherent

File: src/nplinker/metabolomics/abc.py [hunk 6]
```diff
@@ -1,21 +1,27 @@
...
+from .spectrum import Spectrum
 
 
 class SpectrumLoaderBase(ABC):
+    """Abstract base class for SpectrumLoader."""
+
     @property
     @abstractmethod
-    def spectra(self) -> Sequence["Spectrum"]: ...
+    def spectra(self) -> Sequence["Spectrum"]:
```
Comment (gcroci2) [inline src/nplinker/metabolomics/abc.py]: Can we return something more specific here? Again, for coherence, unless it causes other issues

File: src/nplinker/metabolomics/abc.py [hunk 7]
```diff
@@ -26,17 +32,31 @@ def get_mfs(self, keep_singleton: bool) -> Sequence["MolecularFamily"]:
...
         """
 
 
 class FileMappingLoaderBase(ABC):
+    """Abstract base class for FileMappingLoader."""
+
     @property
     @abstractmethod
-    def mappings(self) -> dict[str, list[str]]: ...
+    def mappings(self) -> Mapping[str, Sequence[str]]:
```
Comment (gcroci2) [inline src/nplinker/metabolomics/abc.py]: Maybe in abstract methods, you intentionally want to be more general in the returns type hints. Is this the case?

File: src/nplinker/metabolomics/abc.py [hunk 8]
```diff
@@ -26,17 +32,31 @@ def get_mfs(self, keep_singleton: bool) -> Sequence["MolecularFamily"]:
...
+        """
 
 
 class AnnotationLoaderBase(ABC):
+    """Abstract base class for AnnotationLoader."""
+
     @property
     @abstractmethod
-    def annotations(self) -> dict[str, dict]: ...
+    def annotations(self) -> Mapping[str, Mapping]:
```
Comment (gcroci2) [inline src/nplinker/metabolomics/abc.py]: Same question as above

**Review by gcroci2 — APPROVED** [review]

File: src/nplinker/metabolomics/abc.py (hunk 7 above)
Comment (CunliangGeng) [reply src/nplinker/metabolomics/abc.py]: Indeed, that is intentional. But I thought about it again and agreed that consistence is important. So adding one more principle:
1.  use general types for input and specific type for return
2. only when necessary, use general type for return of abstract methods

It's not necessary here, so I'll change them back to specific types.

**Comment by CunliangGeng** [comment]
### Merge activity

* **Jun 14, 7:27 AM EDT**: @CunliangGeng started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/NPLinker/nplinker/257).
* **Jun 14, 7:36 AM EDT**: [Graphite](https://app.graphite.dev/github/pr/NPLinker/nplinker/257) rebased this pull request as part of a merge.
* **Jun 14, 7:38 AM EDT**: @CunliangGeng merged this pull request with [Graphite](https://app.graphite.dev/github/pr/NPLinker/nplinker/257).

**Pull request merged**